File: DSA/DSA11011 CÂY NHỊ PHÂN ĐỦ.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};



bool check = true;

bool add(TreeNode *node, int parent, int child, char pos)
{
    if (!node) return false;
    
    if (node->val != parent)
    {
        if (add(node->left, parent, child, pos)) return true;
        if (add(node->right, parent, child, pos)) return true;
        return false;
    }
    else
    {
        if (pos == 'L')
        {
            if (!node->left) node->left = new TreeNode(child);
        }
        else
        {
            if (!node->right) node->right = new TreeNode(child);
        }
        return true;
    }
}

void lvl_traversal(TreeNode *node, int depth)
{
    if (!node || (node->left == nullptr && node->right == nullptr))
    {
        return;
    }
    

    if (!node->left || !node->right)
    {
        check = false;
        return;
    }
    
    lvl_traversal(node->left, depth+1);
    lvl_traversal(node->right, depth+1);
}

void submit()
{
    check = true;

    int n;
    cin >> n;

    int parent, child;
    char pos;
    cin >> parent >> child >> pos;

    TreeNode *root = new TreeNode(parent);
    
    add(root, parent, child, pos);

    for (int i = 1; i < n; i++)
    {
        cin >> parent >> child >> pos;
        add(root, parent, child, pos);
    }
    
    lvl_traversal(root, 0);

    if (check)
    {
        cout << 1;
    }
    else
    {
        cout << 0;
    }
    

    cout << endl;
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int so_bo_test;
    cin >> so_bo_test;
    cin.ignore();
    
    while (so_bo_test--)
    {
        submit();
    }

    return 0;
}

File: DSA_python/DSA11011 CÂY NHỊ PHÂN ĐỦ.py
from collections import deque

class TreeNode:
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right


def add_to_tree(edges):
    if not edges:
        return

    nodes = dict()

    for parent, child, pos in edges:
        if parent not in nodes:
            nodes[parent] = TreeNode(parent)
        if child not in nodes:
            nodes[child] = TreeNode(child)
        if pos == 'L':
            nodes[parent].left = nodes[child]
        else:
            nodes[parent].right = nodes[child]
        
    children = [child for _,child,_ in edges]
    parents = [parent for parent,_,_ in edges]
    root = parents[0]

    for parent in parents:
        if parent not in children:
            root = parent
    
    return nodes[root]

def traversal(node):
    if not node.left and not node.right:
        return
    if not node.left or not node.right:
        global check
        check = False
        return

    traversal(node.left)
    traversal(node.right)

def submit():
    n = int(input())
    A = list(map(str, input().split()))
    edges = list()

    for i in range(2, len(A), 3):
        edges.append((A[i-2], A[i-1], A[i]))

    root = add_to_tree(edges)
    global check
    check = True
    traversal(root)

    print(1 if check == True else 0)







test = int(input())
for i in range(test):
    submit()